docs/config-validation.md

```diff
@@ -0,0 +1,191 @@
+# Configuration Validation
+
+We use a fork of
+[github.com/go-playground/validator](https://github.com/go-playground/validator)
+to enable validation of our configuration files using struct tags. The validator
+is configured to validate the configuration file at component startup if the
+`-validate` command line flag is set. If the supplied configuration file is
+valid, the component will exit with a zero exit code, otherwise it will exit
+with a non-zero exit code and print the validation errors to stderr.
+
+## Struct Tag Tips
+
+You can find the full list of struct tags supported by the validator [here]
+(https://pkg.go.dev/github.com/go-playground/validator/v10#section-documentation).
+The following are some tips for struct tags that are commonly used in our
+configuration files.
+
+### `required`
+
+This only validates that the value is not the data types default zero value. For
+numbers ensures value is not zero. For strings ensures value is not "". For
+slices, maps, pointers, interfaces, channels and functions ensures the value is
+not nil. Note, this does not validate slices, maps, etc., as they are not nil.
+For these fields you should use `min=1` or `gte=1` to validate the supplied
+values are not empty.
+
+### `omitempty`
+
+The validator will always run any present validations unless the field is also
+tagged with `omitempty`. Said another way, if a validation is present, and that
+validation cannot be satisfied by the zero-value of that field type, then that
+field is also (technically) required. However, when `omitempty` is present,
+validations will only run when the condition of a `required_` (`with`,
+`with_all`, `unless`, etc.) tag is met.
+
+### `-`
```

Comment:
Question: is `validate="-"` equivalent to not having a validate tag at all? It sounds to me like it should be.

In which case, to me the obvious purpose for `validate="-"` is so that there can be a lint which enforces that every config field has a validation tag. But we don't have such a lint. Would one be hard to find or create?

If it would be hard to create it, then imo we shouldn't use `validate="-"` at all, and should instead just leave those fields blank.

Reply:
Okay, I think I've realized that `validate="-"` is _not_ equivalent to having to validate tag at all. Having no tag means "recurse into this struct, validate its fields", while `"-"` means "stop here, do not recurse into this struct". Therefore my suggestion for the docs for this tag is:

> Normally, config validation descends into all struct-type fields, recursively validating their fields all the way down. Sometimes this can pose a problem, when a nested struct declares one of its fields as required, but a parent struct wants to treat the whole nested struct as optional. The `"-"` tag tells the validation not to recurse, marking the tagged field as optional, and therefore making all of its sub-fields optional as well. We use this tag for many config duration and password file struct valued fields which are optional in some configs but required in others.